### ISSUE TITLE
ctx.coverage_instrumented returns False for targets

### ISSUE THREAD

**Issue by jayconrod** [bazelbuild/bazel#3434]
### Description of the problem / feature request / question:

In Skylark rules, `ctx.coverage_instrumented()` always seems to return False when applied to a target.

We'd like to use this for the Go rules. Basically, we'd like to add coverage instrumentation to sources supplied through a `library` attribute to `go_binary` and `go_test`. The `library` attribute should always be a `go_library` rule (or something compatible). The `srcs` of `go_binary` and `go_test` should not be instrumented.

What we've noticed is that `ctx.coverage_instrumented()` returns True for `go_library` rules when given no argument, but `ctx.coverage_instrumented(ctx.attr.library)` returns False for the same library when called from another rule.

### If possible, provide a minimal example to reproduce the problem:

**def.bzl**
```bzl
def _foo_library_impl(ctx):
  print("%s coverage: %s" % (ctx.label.name, ctx.coverage_instrumented()))
  for d in ctx.attr.deps:
    print("  dep %s coverage: %s" % (d.label.name, ctx.coverage_instrumented(d)))
  f = ctx.new_file("%s.txt" % ctx.label.name)
  ctx.file_action(f, "")

foo_library = rule(
    implementation = _foo_library_impl,
    attrs = {
        "deps": attr.label_list(),
    },
)

def _foo_test_impl(ctx):
  print("%s coverage: %s" % (ctx.label.name, ctx.coverage_instrumented()))
  for d in ctx.attr.deps:
    print("  dep %s coverage: %s" % (d.label.name, ctx.coverage_instrumented(d)))
  ctx.file_action(ctx.outputs.executable, "", True)

foo_test = rule(
    implementation = _foo_test_impl,
    test = True,
    attrs = {
        "deps": attr.label_list(),
    },
)
```

**BUILD.bazel**
```bzl
load("//:def.bzl", "foo_library", "foo_test")

foo_library(
    name = "a_lib",
)

foo_library(
    name = "b_lib",
    deps = [":a_lib"],
)

foo_test(
    name = "foo_test",
    deps = [":b_lib"],
)
```

**Output:**
```
$ bazel coverage :foo_test
____Loading complete.  Analyzing...
WARNING: /Users/jayconrod/Code/scratch/def.bzl:2:3: a_lib coverage: True.
WARNING: /Users/jayconrod/Code/scratch/def.bzl:2:3: b_lib coverage: True.
WARNING: /Users/jayconrod/Code/scratch/def.bzl:4:5:   dep a_lib coverage: False.
WARNING: /Users/jayconrod/Code/scratch/def.bzl:16:3: foo_test coverage: False.
WARNING: /Users/jayconrod/Code/scratch/def.bzl:18:5:   dep b_lib coverage: False.
____Found 1 test target...
blaze: Entering directory `/private/var/tmp/_bazel_jayconrod/8ab5eebb01fab57e02d20efb18e0c6ea/execroot/__main__/'
blaze: Leaving directory `/private/var/tmp/_bazel_jayconrod/8ab5eebb01fab57e02d20efb18e0c6ea/execroot/__main__/'
Target //:foo_test up-to-date:
  bazel-bin/foo_test
____Elapsed time: 0.340s, Critical Path: 0.00s
//:foo_test                                                     (cached) PASSED in 0.1s
```

### Environment info
* Operating System: macOS 10.12.5

* Bazel version (output of `bazel info release`):

Build label: 
Build target: bazel-out/darwin_x86_64-fastbuild/bin/src/main/java/com/google/devtools/build/lib/bazel/BazelServer_deploy.jar
Build time: Fri Jul 21 07:25:44 2017 (1500621944)
Build timestamp: 1500621944
Build timestamp as int: 1500621944

Also observed on 0.5.2.

* If `bazel info release` returns "development version" or "(@non-git)", please tell us what source tree you compiled Bazel from; git commit hash is appreciated (`git rev-parse HEAD`):

This was from Jenkins build 1669. Also observed on 0.5.2.

**Comment by iirina** [comment]
Hi Jay,

This check doesn't work with custom Skylark rules. Currently `coverage_instrumented()` checks if code coverage instrumentation should be generated overall and in your case it returns `True` since you invoked `bazel coverage`. When applied to a target `coverage_instrumented(target)`, bazel will do some internal checks that are currently not exposed to Skylark (i.e. checks if `InstrumentedFilesProvider` is returned by the target - I'm not sure if you're interested in the details) and your rules cannot make use of this.

Long story short: each rule implementation must specify if it should be instrumented or not. In our native rules we rely on a returned provider, but since you are writing your custom rules I suggest to add this logic to your rule implementation. For example, you can make your rule implementation return a declared provider or a field in a struct to mark if the target should be instrumented or not.

WDYT?

**Comment by jayconrod** [comment]
Thanks, that clarifies things a bit. Maybe this is actually a documentation problem then?

> Returns whether code coverage instrumentation should be generated when performing compilation actions for this rule or, if target is provided, the rule specified by that Target.

Reading that, I would expect `ctx.coverage_instrumented` to return the same value whether it was invoked in a rule without an argument, or invoked on a target from another rule. Will that change in the future?

**Comment by iirina** [comment]
> Thanks, that clarifies things a bit. Maybe this is actually a documentation problem then?

You are right, we should update the documentation to be more clear about this case.

> Reading that, I would expect ctx.coverage_instrumented to return the same value whether it was invoked in a rule without an argument, or invoked on a target from another rule. Will that change in the future?

No, it shouldn't change in the future. There may be use cases when you don't want certain code to be instrumented (for example auto-generated code). Bazel has no way of knowing whether a custom target should be instrumented or not (that's up to its owner). If you don't want to exclude any code from being instrumented than you can simply rely on the value of `ctx.coverage_instrumented` without arguments.

**Comment by iirina** [comment]
Also test targets are not instrumented by default, so if you would call `ctx.coverage_instrumented(my_java_test_target)` it would return `False` (this is just FYI regarding to why for some targets `ctx.coverage_instrumented` returns a different value).

**Comment by jayconrod** [comment]
Right, it makes sense that test targets are not instrumented. It would be good for that to be documented, too.

In our case, if we called `ctx.coverage_instrumented(target)` from a `go_test` rule on a `go_library` target, it would return False, even though `ctx.coverage_instrumented()` returned true within the same `go_library`.

Go has some idiosyncrasies that sometimes require library sources to be compiled by test rules instead of the libraries themselves, so this required a workaround.

**Comment by iirina** [comment]
I see. Other than updating the documentation is there something else I can do? How do you plan to approach this?

**Comment by jayconrod** [comment]
No other action needed from Bazel, aside from the documentation change I think.

As you suggested, returning something through a provider that indicates coverage was requested seems like the best way to proceed on our side.
